### DISCUSSION TITLE
[Feature] Add option to use embedded preview for RAW images

### DISCUSSION THREAD

**Discussion by suryanarayanan-v** [immich-app/immich#5758] [Feature Request]
### The feature

Currently, Immich generates thumbnails for RAW images (DNG, CR2, etc.) using the Sharp library, which processes the entire RAW file. This leads to thumbnails with poor contrast and dimmed highlights. To improve this, Immich should extract the embedded JPEG previews in RAW files using ExifTool, providing more accurate and visually representative thumbnails.

Proposed Change:

Switch thumbnail generation for RAW images to use embedded JPEG previews.
Extract JPEG previews with ExifTool command like: exiftool -b -PreviewImage -w _preview.jpg /path_to_raw_file

For example heres how the image looks in immich:
<img width="621" alt="Screenshot 2023-12-17 at 12 34 15 AM" src="https://github.com/immich-app/immich/assets/41332748/60f17361-750f-4082-b487-63a1744ae162">

and heres how the image should actually look like:
<img width="591" alt="Screenshot 2023-12-17 at 12 34 35 AM" src="https://github.com/immich-app/immich/assets/41332748/c04300e8-e09d-4279-bcfd-0a287a4936f8">

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mertalev** [comment]
This is a good idea, but this would be in addition to the current approach, not a replacement. Many RAW images are rendered well with the sharp pipeline and many don't have high resolution embedded previews. But it's true that in some cases (like with your camera), there is an embedded preview more in line with user expectations, so adding a setting to prefer embedded previews if available would be fine.

**Reply by suryanarayanan-v** [reply]
Unfortunately this is how all of my images look. This is just an image from my iPhone with the proRaw feature enabled. So its not like its a niche camera.

**Comment by stevenla** [comment]
it would be pretty nice to be able to define exactly how thumbnails are extracted on a per-library basis. For example, my camera (Sony a6000) has abysmally small thumbnails (1024px wide), but has an option to shoot ARW+JPEG. The JPEGs are totally usable as thumbnails and have the in-camera color correction and lens distortion correction baked in.

I'd personally prefer if my immich library for my camera raws were pointers to the actual ARW files (so pressing download would download the raw), but the thumbnails were generated from the JPEG files.

A setting to change the shell command of the thumbnail extractor could be flexible enough to be adaptable to a variety of use cases.

**Reply by mertalev** [reply]
> I'd personally prefer if my immich library for my camera raws were pointers to the actual ARW files (so pressing download would download the raw), but the thumbnails were generated from the JPEG files.

This was actually added recently. There's an option to use the embedded JPEG as the input for thumbnail generation instead of the RAW.

**Reply by mertalev** [reply]
Ah, re-reading this I think you might mean having this interaction with separate JPEG and ARW files. You're right that this isn't currently possible.

We do plan to have auto-stacking for JPEG+RAW pairs in the future, so while the thumbnail of the RAW asset would not be derived from the JPEG, the displayed asset would ultimately be the JPEG.

**Reply by stevenla** [reply]
that would also work to suit my use case. thanks for the update!

**Comment by 2001Y** [comment]
I would very much like this feature.
The current 'low contrast/darkness of RAW previews' problem,
I think the problem can be solved by the following steps/conditions.

1. use a high quality embedded preview JPEG of the RAW (useful for many cameras including iPhone)
#5758
2. if there is no RAW high quality embedded preview JPEG, but there is a JPEG with the same name/meta information, "Stack" it. 
#2479 
3. if none are available, use the current RAW to JPEG conversion using Sharp.

We think that the current "When shooting RAW with iPhone, only the RAW is backed up and previewing is done using Sharp's RAW to JPEG conversion" is not good anymore.

**Comment by AngelaDMerkel** [comment]
@mertalev This issue would be partially resolved by passing libraw the correct icc profiles or by designing one which fit the ProRAWs rendering intent. In this way it would be possible to produce a full resolution image from a "RAW" file rather than having to extract a lower resolution embedded preview. 

I would be willing to make a few ICC profiles for the most popular cameras, although Adobe has done most of the legwork: 

[Apple iPhone13,3 back camera Adobe Standard.dcp.zip](https://github.com/immich-app/immich/files/15250826/Apple.iPhone13.3.back.camera.Adobe.Standard.dcp.zip)

**Reply by mertalev** [reply]
That might get closer to the DNG preview, but I think the root cause for the darkness is that the DNG has a `ProfileGainTableMap` field that libraw ignores. The RAW image assumes it'll be interpreted with this gain map, so the final brightness looks off when it isn't used.

**Reply by AngelaDMerkel** [reply]
Yes, it wouldn't produce an identical image to the preview, but it's easily possible to make an ICC profile that looks good and not overly dark. It's a lot of work, however.

**Comment by bo0tzz** [comment]
This option is now available.
